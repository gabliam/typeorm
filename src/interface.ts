import { interfaces } from '@gabliam/core';

export interface TypeormRegistry extends interfaces.ValueRegistry {
  filePath: string;
}
