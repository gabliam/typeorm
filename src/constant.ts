export const ConnectionOptionsBeanId = Symbol('ConnectionOptionsBeanId');

export const TYPE = {
  Entity: 'EntityTypeOrm'
};

export const ENTITIES_PATH = Symbol('TYPEORM/ENTITIES_PATH');
