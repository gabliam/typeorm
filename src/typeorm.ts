export * from 'typeorm';

import {
  Entity,
  AbstractEntity,
  ClassEntityChild,
  SingleEntityChild,
  ClosureEntity,
  EmbeddableEntity
} from './decorators';

export {
  Entity,
  AbstractEntity,
  ClassEntityChild,
  SingleEntityChild,
  ClosureEntity,
  EmbeddableEntity
};
