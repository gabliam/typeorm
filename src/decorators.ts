import {
  Entity as typeormEntity,
  AbstractEntity as typeormAbstractEntity,
  ClassEntityChild as typeormClassEntityChild,
  SingleEntityChild as typeormSingleEntityChild,
  ClosureEntity as typeormClosureEntity,
  EmbeddableEntity as typeormEmbeddableEntity,
  EntityOptions
} from 'typeorm';
import { register } from '@gabliam/core';
import { TYPE } from './constant';

/**
 * This decorator is used to mark classes that will be an entity (table or document depend on database type).
 * Database schema will be created for all classes decorated with it, and Repository can be retrieved and used for it.
 */
export function Entity(name?: string, options?: EntityOptions) {
  return (target: Function) => {
    register(TYPE.Entity, { id: target, target })(target);
    typeormEntity(name, options)(target);
  };
}

/**
 * Abstract entity is a class that contains columns and relations for all entities that will inherit this entity.
 * Database table for the abstract entity is not created.
 */
export function AbstractEntity() {
  return (target: Function) => {
    register(TYPE.Entity, { id: target, target })(target);
    typeormAbstractEntity()(target);
  };
}

/**
 * Special type of the entity used in the class-table inherited tables.
 */
export function ClassEntityChild(tableName?: string, options?: EntityOptions) {
  return (target: Function) => {
    register(TYPE.Entity, { id: target, target })(target);
    typeormClassEntityChild(tableName, options)(target);
  };
}

/**
 * Special type of the table used in the single-table inherited tables.
 */
export function SingleEntityChild() {
  return (target: Function) => {
    register(TYPE.Entity, { id: target, target })(target);
    typeormSingleEntityChild()(target);
  };
}

/**
 * Used on a entities that stores its children in a tree using closure design pattern.
 */
export function ClosureEntity(name?: string, options?: EntityOptions) {
  return (target: Function) => {
    register(TYPE.Entity, { id: target, target })(target);
    typeormClosureEntity(name, options)(target);
  };
}

/**
 * This decorator is used on the entities that must be embedded into another entities.
 */
export function EmbeddableEntity() {
  return (target: Function) => {
    register(TYPE.Entity, { id: target, target })(target);
    typeormEmbeddableEntity()(target);
  };
}
