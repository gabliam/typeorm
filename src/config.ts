import { PluginConfig, Value, optional, Bean, inject } from '@gabliam/core';
import { ConnectionOptions, Connection, createConnection } from './typeorm';
import { ConnectionOptionsBeanId, ENTITIES_PATH } from './constant';
import * as d from 'debug';

const debug = d('Gabliam:Plugin:Typeorm');

@PluginConfig()
export class PluginTypeormConfig {
  @Value('application.typeorm.connectionOptions')
  connectionOptions: ConnectionOptions;

  entitiesPath: string[];

  constructor(
    @inject(ConnectionOptionsBeanId)
    @optional()
    connectionOptions: ConnectionOptions,
    @inject(ENTITIES_PATH) entitiesPath: string[]
  ) {
    debug('constructor PluginTypeormConfig', connectionOptions);
    this.connectionOptions = connectionOptions;
    this.entitiesPath = entitiesPath;
  }

  @Bean(Connection)
  create() {
    debug('connectionOptions', this.connectionOptions);
    if (!this.connectionOptions) {
      throw new Error(`PluginTypeormConfig connectionOptions is mandatory`);
    }
    const connectionOptions = this.connectionOptions;
    const entities: any = connectionOptions.entities || [];
    entities.push(...this.entitiesPath);

    return createConnection({
      ...connectionOptions,
      entities
    });
  }
}
