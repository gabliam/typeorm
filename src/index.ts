import { interfaces, Scan, inversifyInterfaces, Registry } from '@gabliam/core';
import { Connection } from './typeorm';

export * from './typeorm';
export * from './constant';


@Scan(__dirname)
export default class PluginsTypeOrm implements interfaces.GabliamPlugin {
    async destroy(container: inversifyInterfaces.Container, registry: Registry) {
        let connection = container.get<Connection>(Connection);
        await connection.close();
    }
}
