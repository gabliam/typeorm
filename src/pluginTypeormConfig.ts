import { PluginConfig, Value, optional, Bean, CORE_CONFIG, inject, interfaces } from '@gabliam/core';
import { ConnectionOptions, Connection, createConnection } from './typeorm';
import {ConnectionOptionsBeanId} from './constant';
import * as d from 'debug';

const debug = d('Gabliam:Plugin:Typeorm');

@PluginConfig()
export class PluginTypeormConfig {

    @Value('application.typeorm.connectionOptions')
    connectionOptions: ConnectionOptions;

    @Value('application.typeorm.entitiesPath')
    entitiesPath: string;

    constructor(
        @inject(ConnectionOptionsBeanId) @optional() connectionOptions: ConnectionOptions,
        @inject(CORE_CONFIG) config: interfaces.GabliamConfig
    ) {
        debug('constructor PluginTypeormConfig', connectionOptions, config);
        this.connectionOptions = connectionOptions;
        this.entitiesPath = `${config.discoverPath}/**/*{.js,.ts}`;
    }

    @Bean(Connection)
    create() {
        debug('connectionOptions', this.connectionOptions);
        if (!this.connectionOptions){
            throw new Error(`PluginTypeormConfig connectionOptions is mandatory`);
        } 
        let connectionOptions = this.connectionOptions;
        let entities: any = connectionOptions.entities || [];
        entities.push(this.entitiesPath);

        return createConnection({
            ...connectionOptions,
            entities
        });
    }
}