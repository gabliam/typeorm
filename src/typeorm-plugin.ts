import { interfaces, Scan, inversifyInterfaces, Registry } from '@gabliam/core';
import { Connection } from './typeorm';
import { TYPE, ENTITIES_PATH } from './constant';
import { TypeormRegistry } from './interface';

@Scan(__dirname)
export class TypeOrmPlugin implements interfaces.GabliamPlugin {
  bind(container: inversifyInterfaces.Container, registry: Registry) {
    const entitiesPaths = registry
      .get<TypeormRegistry>(TYPE.Entity)
      .map(({ target }) => {
        return target;
      });
    container.bind<any>(ENTITIES_PATH).toConstantValue(entitiesPaths);
  }

  async destroy(container: inversifyInterfaces.Container, registry: Registry) {
    const connection = container.get<Connection>(Connection);
    await connection.close();
  }
}
