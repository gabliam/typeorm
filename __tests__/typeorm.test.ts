import { TypeormPluginTest } from './typeorm-plugin-test';
import { Entity, Column, PrimaryGeneratedColumn, Connection } from '../src';
import { Gabliam } from '@gabliam/core';

let appTest: TypeormPluginTest;
beforeEach(async () => {
  appTest = new TypeormPluginTest();
});

afterEach(async () => {
  try {
    await appTest.destroy();
  } catch (e) {}
});

test('without config folder', async () => {
  await expect(appTest.gab.buildAndStart()).rejects.toMatchSnapshot();
});

test('with bad config', async () => {
  appTest.addConf('application.typeorm.connectionOptions', 'bad');
  await expect(appTest.gab.buildAndStart()).rejects.toMatchSnapshot();
});

test('with no driver', async () => {
  appTest.addConf('application.typeorm.connectionOptions', {
    driver: {
      type: 'postgres',
      host: 'localhost',
      username: 'test',
      password: 'test',
      database: 'test'
    },
    autoSchemaSync: true
  });
  await expect(appTest.gab.buildAndStart()).rejects.toMatchSnapshot();
});

test('with config', async () => {
  appTest.addConf('application.typeorm.connectionOptions', {
    driver: {
      type: 'sqlite',
      storage: 'test.sqlite'
    },
    autoSchemaSync: true
  });

  @Entity()
  class Photo {
    @PrimaryGeneratedColumn() id?: number;

    @Column({
      length: 500
    })
    name: string;

    @Column('text') description: string;

    @Column() fileName: string;

    @Column('int') views: number;

    @Column() isPublished: boolean;
  }
  appTest.addClass(Photo);

  await expect(appTest.gab.buildAndStart()).resolves.toBeInstanceOf(Gabliam);
  const connection = appTest.gab.container.get(Connection);
  expect(connection.isConnected).toMatchSnapshot();
  const repo = connection.getRepository<Photo>('Photo');
  let res = await repo.find();
  expect(res).toMatchSnapshot();
  await repo.persist({
    name: 'test',
    description: 'test desc',
    fileName: 'testfile',
    views: 0,
    isPublished: false
  });
  res = await repo.find();
  expect(res).toMatchSnapshot();
  await connection.dropDatabase();
});
