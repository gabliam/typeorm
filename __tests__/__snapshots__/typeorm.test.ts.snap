// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with bad config 1`] = `[TypeError: Cannot read property 'type' of undefined]`;

exports[`with config 1`] = `true`;

exports[`with config 2`] = `Array []`;

exports[`with config 3`] = `
Array [
  Photo {
    "description": "test desc",
    "fileName": "testfile",
    "id": 1,
    "isPublished": false,
    "name": "test",
    "views": 0,
  },
]
`;

exports[`with no driver 1`] = `[DriverPackageNotInstalledError: Postgres package has not been found installed. Try to install it: npm install pg --save]`;

exports[`without config folder 1`] = `[Error: PluginTypeormConfig connectionOptions is mandatory]`;
